File: 20. Valid Parentheses.py
# 20. Valid Parentheses
# Time: O(N)
# Space: O(N)

# use stack and map to match pairs.
class Solution:
    def isValid(self, s: 'str') -> 'bool':
        stack = []
        p = {'{': '}', '[': ']', '(': ')'}
        for c in s:
            if c in p:
                stack.append(p[c])
            elif len(stack) > 0 and c == stack[-1]:
                stack.pop()
            else:
                return False

        return len(stack) == 0

File: 20. Valid Parentheses.java
// 20. Valid Parentheses
// Time: O(N)
// Space: O(N)

// pair corresponding brackets using stack
public class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            if (c == '(') {
                stack.push(')');
            } else if (c == '[') {
                stack.push(']');
            } eles if (c == '{') {
                stack.push('}');    
            }else if (stack.isEmpty() || stack.pop() != c) {
                return false;
            }
        }
        return stack.isEmpty();
    }
}